RFC: Move log writes to thread.

zwn opened the pull request:
The basics work. It needs to handle finishing, thread joining, KeyboardInterupt etc.

m3d commented:
OK, can you please squash both commits so that I can cherry-pick it for K2 testing? (yes, I can pick both, but you will squash it later anyway ;) ...
I email you wrote that it needs finishing, but here it looks like complete?? What about `self.setDaemon(True)`?

zwn commented:
> OK, can you please squash both commits so that I can cherry-pick it for K2 testing? (yes, I can pick > both, but you will squash it later anyway ;) ...

Will do.

> I email you wrote that it needs finishing, but here it looks like complete??

Yes, I have been finishing it in the meantime.

> What about self.setDaemon(True)?

No!!! That would defeat the whole purpose of this exercise. We never want the writing thread to be killed without being able to finish the writing. It needs to end properly otherwise we will be missing data in the log.

zwn commented:
Rebased on current master.

m3d commented on subt/main.py:
```diff
@@ -459,11 +458,12 @@ def main():
-        game = robot.modules['app']  # TODO nicer reference
-        robot.start()
-        game.play()
+        with LogWriter(prefix=prefix, note=str(sys.argv)) as log:
```
we should really use the `launcher`

zwn replied:
first things first

zwn commented:
I have gone over all usages of `LogWriter` and converted them to context manager to ensure proper waiting for the writer thread.

m3d commented:
OK, thanks - Jakub should test it tomorrow morning. I would maybe test also `Process` if this test fails.

zwn commented:
I have rebased this branch (feature/writer) on current master. Changes to use LogWriter as context manager were already in, so this branch is now simplified.

Anyway, the problem seems to be with calling `write` with too small buffer which significantly limits throughput. Here is a benchmark:

```python
#!/usr/bin/python3

import numpy
import sys
import time
import os

BLOCK_COUNT = 1000

print('block_size [B] | rate [MB/s]')
for block_size in [2**n for n in range(1, 26)]:
    block = numpy.random.bytes(block_size)
    f = os.open(sys.argv[1], os.O_WRONLY)
    start_time = time.time()
    for i in range(BLOCK_COUNT):
        os.write(f, block)
    os.close(f)
    end_time = time.time()
    duration = end_time - start_time
    volume = block_size * BLOCK_COUNT
    rate_bytes_per_second = volume / duration
    rate_megabytes_per_second = rate_bytes_per_second / 2**20
    print(block_size, rate_megabytes_per_second)
```

On my fairly new and powerful notebook I get this output:

```
block_size [B] | rate [MB/s]
2 0.2390914524805738
4 3.341687552213868
8 5.704099821746881
16 10.718472617651985
32 18.85402857563706
64 47.67225325884544
128 84.155161078238
256 146.20217018846373
512 220.78482104355325
1024 326.42652215492507
2048 435.74468085106383
4096 2320.022656471255
8192 2732.032683008171
16384 2201.3368714520843
32768 2942.066395816031
65536 3577.9273069731257
131072 3759.33401690771
262144 3775.400645930172
524288 1077.683697836721
1048576 2823.9736232779824
2097152 2733.161388194172
4194304 1293.071493516498
8388608 787.1373734788087
16777216 818.7598566093095
33554432 823.4487270467835
```

This seems to suggest that we should be using write buffer between the sizes 64kK až 256kB. For a reference, gnu cp uses 128kB (https://eklitzke.org/efficient-file-copying-on-linux).

zwn commented:
I tried also this benchmark:

```python
def benchmark(filepath):
    with LogReader(filepath) as logIn:
        with LogWriter() as logOut:
            for timestamp, stream_id, data in logIn:
                logOut.write(stream_id, data)
```

with a 617MB logfile from Robik. It took 4.6s so the throughput was 134MB/s which is significantly less than the maximum which is around 3GB/s. So it seems that indeed the write buffer size has a significant effect.
